### ISSUE TITLE
AccessToken does not include email claim even when Email scope is requested

### ISSUE THREAD

**Issue by akamud** [IdentityServer/IdentityServer4#678]
- [x] I read and understood how to enable [logging](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging)

### Issue / Steps to reproduce the problem
I'm requesting the `email` scope when I'm authenticating, the `email` is returned form the User Info endpoint, but it is not included in the `acess_token`.
Is this expected? *Should* it be included in the `access_token`?

I can see from [this code](https://github.com/IdentityServer/IdentityServer4/blob/62eea365e8835323230c32eff6ec9b7b78cc1be0/src/IdentityServer4/Services/DefaultClaimsService.cs#L145) that the `email` claim will never be included unless I have a Resource Scope and I set it to `IncludeAllClaimsForUser`.

I think it is a bit weird that I explicitly request an `email` scope and it can only be included in the `access_token` if **another** scope is used and I have no way to customize this.

The relevant code is here:

ScopeStore implementation:
```C#
public async Task<IEnumerable<Scope>> FindScopesAsync(IEnumerable<string> scopeNames) =>
            (await GetScopesAsync()).Where(x => scopeNames.Contains(x.Name));

public Task<IEnumerable<Scope>> GetScopesAsync(bool publicOnly = true)
        {
            var list = new List<Scope>
            {
                new Scope
                {
                    Name = StandardScopes.OpenId.Name,
                    Required = true,
                    Type = (int) ScopeType.Identity,
                    Enabled = true
                },
                new Scope
                {
                    Name = StandardScopes.Profile.Name,
                    Type = (int) ScopeType.Identity,
                    Emphasize = true,
                    Enabled = true
                },
                new Scope
                {
                    Name = StandardScopes.Email.Name,
                    Type = (int) ScopeType.Identity,
                    Emphasize = true,
                    Enabled = true
                },
                new Scope
                {
                    Name = StandardScopes.OfflineAccess.Name,
                    Type = (int) ScopeType.Identity,
                    Emphasize = true,
                    Enabled = true
                }
            };

            return Task.FromResult(list.AsEnumerable());
        }
```

Then in my `IProfileService` I have implemented in my `GetProfileDataAsync` (simplified here) a method that returns the user email and some other info.

```
public async Task GetProfileDataAsync(ProfileDataRequestContext context)
        {
            var user = await people.FindByLoginOption(context.Subject.GetSubjectId());

            var claims = new List<Claim> { new Claim(Constants.ClaimTypes.Email, user.Email) };
            context.IssuedClaims.AddRange(claims);

            await Task.FromResult(0);
        }
```

### Relevant parts of the log file

I can see that the `email` scope is correctly recognized by IS4
```
 Authorize request validation success
{
        "ClientId": "MyClientId",
        "ClientName": "Client Name",
        "RedirectUri": "http://localhost:5000/signin-oidc",
        "AllowedRedirectUris": [
          "hidden"
        ],
        "SubjectId": "foo",
        "ResponseType": "code",
        "ResponseMode": "form_post",
        "GrantType": "authorization_code",
        "RequestedScopes": "openid profile email offline_access",
...
```

Later I see this in the logs:
```
Scopes in access token: openid profile email offline_access
```

Then this, when it tries to put all the user claims in the `access_token`, which is the code I linked above:
```
Getting claims for access token for subject: foo
```

Seeing the code, I think the real question here is: is my assumption correct that the `email` claim should be in the `access_token`? If not, shouldn't this be something we could control? Right now the code does not allow for this unless I use a Resource Scope and then use `IncludeAllClaimsForUser`.

**Comment by vquaiato** [comment]
@brockallen Is this information present in the docs or I should have known for knowing the OAuth/OpenId process and flow? If it is in the doc could you please link it?

**Comment by brockallen** [comment]
This is what we have: https://identityserver4.readthedocs.io/en/release/configuration/resources.html

If it's not clear, we accept PRs.

**Comment by akamud** [comment]
I read the docs but I still have some questions:
When you say "configure the claim in both" you mean that `access_token` information only come from Resource Scopes?
There is no way to add info to the `access_token` without adding a resource scope?

**Comment by brockallen** [comment]
> When you say "configure the claim in both" you mean that access_token information only come from Resource Scopes?

access_token claims come from ApiResources, and id_token claims come from IdentityResources.

**Comment by akamud** [comment]
Yes, I see that from the code I linked, but is this an OpenID/OAuth restriction or is it a IS limitation?

**Comment by brockallen** [comment]
This is by design.

**Comment by akamud** [comment]
Please can you be more clear? What is the motivation of this decision? This is an OpenID design or IS design?

This is an open source software, people must be certain when they see something they can help improve before they send a PR.
Right now I'm not sure if this is something that the community can help improve or if doing this would break some OpenID specification.

**Comment by brockallen** [comment]
This is an IS design. You get two tokens and you need a way to control what's in each. The IdentityResource and ApiResource configuration object mode is how that's expressed. 

What needs changing?

**Comment by akamud** [comment]
Ok, now the decision is clear ;)

Do you think it makes sense to have another way to add claims to `id_token` or `access_token` that does not tie Claims with Scopes?

**Comment by brockallen** [comment]
Scopes are resources the client is requesting access to. For id_token, those are absolutely tied. For access_tokens we felt it it was a good place to allow an API to express what info it needed about the user.

What other approach do you have in mind?

**Comment by akamud** [comment]
We have a scenario where one of our clients is a SaaS application that can't be customized, but it implements OAuth. The problem is that it tries to retrieve some claims (e.g. `email`) from the `access_token`, if it doesn't find it there, it throws an error, failing the authentication process. This client always sends the `openid profile email` scopes when authenticating.

So in our scenario, telling this client to request the `api` scope is not an option, but we think it makes sense to add the `email` claim inside the `access_token` since he is requesting this scope. I can see from the code that, since the `email` is an Identity Scope, it will not be included in the `access_token`. So we tried to find some way to work around this.

What we tried, with no success, was adding some claims inside the `GetProfileDataAsync`, using this:
```C#
context.IssuedClaims.AddRange(claims);
```
Since I have access to the `ProfileDataRequestContext` and the `IssuedClaims` and I can Add any number of claims to it, we thought it made sense.
But I saw that they only get added to the `access_token` when [there is an `additionalClaims` to be added](https://github.com/IdentityServer/IdentityServer4/blob/62eea365e8835323230c32eff6ec9b7b78cc1be0/src/IdentityServer4/Services/DefaultClaimsService.cs#L183). Which will always be empty since I don't have an API Scope in the request.

My suggestion is having some way to control the `additionalClaims` or the `IssuedClaims` itself, so we could add other Claims to the `access_token` without having to necessarily introduce a new API scope.

Is this [condition](https://github.com/IdentityServer/IdentityServer4/blob/62eea365e8835323230c32eff6ec9b7b78cc1be0/src/IdentityServer4/Services/DefaultClaimsService.cs#L183) really necessary? Removing it would always try to add what is inside `IssuedClaims`, even when no API Resource is used.

**Comment by brockallen** [comment]
> one of our clients is a SaaS application that can't be customized, but it implements OAuth

So the SaaS is a client of your token service? It's getting an access token to call your API?

**Comment by akamud** [comment]
Actually it is not a SaaS, I did not express myself right. It is just an out of the box application that does not allow for customization in the authorization process.

**Comment by akamud** [comment]
Actually it is not a SaaS, I did not express myself right. It is just an application that does not allow for customization in many parts, one of them being the authorization process.

But yes, it gets an `access_token` to call our API.

**Comment by brockallen** [comment]
But if it's the client and it's looking inside the access token, then it's really breaking the rules and I have little sympathy for them.

But yea, maybe we can remove the guard so your custom profile service can add custom claims regardless of the user claims configuration.

**Comment by brockallen** [comment]
opened an issue to track: #683

**Comment by akamud** [comment]
I agree, we also think there are many wrong things with their implementation.

I can help sending a PR to this.

**Comment by brockallen** [comment]
If you do a PR, please include automated tests. If not, we can get to it at some point.

**Comment by akamud** [comment]
As discussed in #683, this was fixed in 1.0.2. We can now add more claims to both the `id_token` and the `access_token` by adding them to the `IssuedClaims` from `ProfileDataRequestContext`.
